Neste projeto, aplicararemos machine learning nos dados extraídos do google analytics com a loja de brindes do Google.

# Imports

In [132]:
# data wrangling
import pandas as pd
import numpy as np
import json

# machine learning
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Get Data

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/alura-cursos/digital-marketing-ml/master/train.csv')

# Carregando a base de dados corretamente

In [3]:
df.head()

channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2      Affiliates  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1  {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2  {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Southea...   
1  {"continent": "Europe", "subContinent": "Easte...   
2  {"continent": "Americas", "subContinent": "Sou...   
3  {"continent": "Americas", "subContinent": "Nor...   
4  {"continent": "Americas", "subContinent": "Nor...   

                        sessionId  socialEngagementType  \
0  9674781571160116268_1472804607  Not Socially Engaged   
1  8590648239310839049_1472835928  Not Socially Engaged   
2  9772828344252850532_1472856802  Not Socially Engaged   
3  1350700416054916432_1472879649  Not Socially Engaged   
4  1350700416054916432_1472829671  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "5", "pageviews": "4",...   

                                       trafficSource     visitId  visitNumber  \
0  {"campaign": "(not set)", "source": "google", ...  1472804607            1   
1  {"campaign": "(not set)", "source": "google", ...  1472835928            1   
2  {"campaign": "Data Share Promo", "source": "Pa...  1472856802            1   
3  {"campaign": "(not set)", "source": "google", ...  1472879649            2   
4  {"campaign": "(not set)", "source": "google", ...  1472829671            1   

   visitStartTime  
0      1472804607  
1      1472835928  
2      1472856802  
3      1472879649  
4      1472829671

In [4]:
df.shape

(12283, 12)

In [5]:
df['fullVisitorId'].unique()

array([9674781571160116268, 8590648239310839049, 9772828344252850532, ...,
        696320405243257438, 5437703324867866371, 2905034275926274980],
      dtype=uint64)

In [6]:
df['fullVisitorId'].nunique()

9996

In [7]:
df.dtypes

channelGrouping         object
date                     int64
device                  object
fullVisitorId           uint64
geoNetwork              object
sessionId               object
socialEngagementType    object
totals                  object
trafficSource           object
visitId                  int64
visitNumber              int64
visitStartTime           int64
dtype: object

Os dados da coluna date foram identificados como inteiro. Não é possível separar os valores da data quando o formato do dado é número inteiro. O campo fullVisitorId está como inteiro também pois pode-se ter perdido zeros a esquerda, será necessário também corrigir a coluna totais, o visitId tem o mesmo problema que o fullVisitorId.

Para resolver isso, utiliza-se a função dype do Pandas ao ler o arquivo.

In [8]:
df = pd.read_csv('https://raw.githubusercontent.com/alura-cursos/digital-marketing-ml/master/train.csv', dtype = {'date':object, 'fullVisitorId':object, 'visitId':object})

In [9]:
df.dtypes

channelGrouping         object
date                    object
device                  object
fullVisitorId           object
geoNetwork              object
sessionId               object
socialEngagementType    object
totals                  object
trafficSource           object
visitId                 object
visitNumber              int64
visitStartTime           int64
dtype: object

In [10]:
df.head()

channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2      Affiliates  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1  {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2  {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Southea...   
1  {"continent": "Europe", "subContinent": "Easte...   
2  {"continent": "Americas", "subContinent": "Sou...   
3  {"continent": "Americas", "subContinent": "Nor...   
4  {"continent": "Americas", "subContinent": "Nor...   

                        sessionId  socialEngagementType  \
0  9674781571160116268_1472804607  Not Socially Engaged   
1  8590648239310839049_1472835928  Not Socially Engaged   
2  9772828344252850532_1472856802  Not Socially Engaged   
3  1350700416054916432_1472879649  Not Socially Engaged   
4  1350700416054916432_1472829671  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "5", "pageviews": "4",...   

                                       trafficSource     visitId  visitNumber  \
0  {"campaign": "(not set)", "source": "google", ...  1472804607            1   
1  {"campaign": "(not set)", "source": "google", ...  1472835928            1   
2  {"campaign": "Data Share Promo", "source": "Pa...  1472856802            1   
3  {"campaign": "(not set)", "source": "google", ...  1472879649            2   
4  {"campaign": "(not set)", "source": "google", ...  1472829671            1   

   visitStartTime  
0      1472804607  
1      1472835928  
2      1472856802  
3      1472879649  
4      1472829671

Algumas colunas possuem dicionários em seus valores e será necessário tratá-los.

In [11]:
df['device'].iloc[0]

'{"browser": "Chrome", "browserVersion": "not available in demo dataset", "browserSize": "not available in demo dataset", "operatingSystem": "Windows", "operatingSystemVersion": "not available in demo dataset", "isMobile": false, "mobileDeviceBranding": "not available in demo dataset", "mobileDeviceModel": "not available in demo dataset", "mobileInputSelector": "not available in demo dataset", "mobileDeviceInfo": "not available in demo dataset", "mobileDeviceMarketingName": "not available in demo dataset", "flashVersion": "not available in demo dataset", "language": "not available in demo dataset", "screenColors": "not available in demo dataset", "screenResolution": "not available in demo dataset", "deviceCategory": "desktop"}'

Para passar estes valores aos modelos de machine learning, será necessário criar colunas únicas para cada uma das chaves e colocar os valores do dicionário nelas.

In [12]:
type(df['device'].iloc[0])

str

O python identifica estes valores como string e será necessário convertê-los para dicionários. Para isso, será utilizada a biblioteca json.

In [13]:
# identifica cada par de chave valor em linhas distintas
json.loads(df['device'].iloc[0])

{'browser': 'Chrome',
 'browserVersion': 'not available in demo dataset',
 'browserSize': 'not available in demo dataset',
 'operatingSystem': 'Windows',
 'operatingSystemVersion': 'not available in demo dataset',
 'isMobile': False,
 'mobileDeviceBranding': 'not available in demo dataset',
 'mobileDeviceModel': 'not available in demo dataset',
 'mobileInputSelector': 'not available in demo dataset',
 'mobileDeviceInfo': 'not available in demo dataset',
 'mobileDeviceMarketingName': 'not available in demo dataset',
 'flashVersion': 'not available in demo dataset',
 'language': 'not available in demo dataset',
 'screenColors': 'not available in demo dataset',
 'screenResolution': 'not available in demo dataset',
 'deviceCategory': 'desktop'}

In [14]:
type(json.loads(df['device'].iloc[0]))

dict

A biblioteca json identificou como dicionários. Será necessário criar uma lista com todos os valores, através de list comprehension.

In [15]:
device = pd.DataFrame([json.loads(linha) for linha in df['device']])

In [16]:
device.head()

browser                 browserVersion                    browserSize  \
0  Chrome  not available in demo dataset  not available in demo dataset   
1  Chrome  not available in demo dataset  not available in demo dataset   
2  Chrome  not available in demo dataset  not available in demo dataset   
3  Safari  not available in demo dataset  not available in demo dataset   
4  Safari  not available in demo dataset  not available in demo dataset   

  operatingSystem         operatingSystemVersion  isMobile  \
0         Windows  not available in demo dataset     False   
1       Macintosh  not available in demo dataset     False   
2       Macintosh  not available in demo dataset     False   
3             iOS  not available in demo dataset      True   
4             iOS  not available in demo dataset      True   

            mobileDeviceBranding              mobileDeviceModel  \
0  not available in demo dataset  not available in demo dataset   
1  not available in demo dataset  not available in demo dataset   
2  not available in demo dataset  not available in demo dataset   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

             mobileInputSelector               mobileDeviceInfo  \
0  not available in demo dataset  not available in demo dataset   
1  not available in demo dataset  not available in demo dataset   
2  not available in demo dataset  not available in demo dataset   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

       mobileDeviceMarketingName                   flashVersion  \
0  not available in demo dataset  not available in demo dataset   
1  not available in demo dataset  not available in demo dataset   
2  not available in demo dataset  not available in demo dataset   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

                        language                   screenColors  \
0  not available in demo dataset  not available in demo dataset   
1  not available in demo dataset  not available in demo dataset   
2  not available in demo dataset  not available in demo dataset   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

                screenResolution deviceCategory  
0  not available in demo dataset        desktop  
1  not available in demo dataset        desktop  
2  not available in demo dataset        desktop  
3  not available in demo dataset         mobile  
4  not available in demo dataset         mobile

Agora a coluna device foi transformada em um dataframe. Agora será necessário juntar estes novos dataframes no dataframe original e droppar as colunas que as geraram.

In [17]:
geoNetwork = pd.DataFrame([json.loads(linha) for linha in df['geoNetwork']])

In [18]:
geoNetwork.head()

continent      subContinent        country                         region  \
0      Asia    Southeast Asia      Singapore                      (not set)   
1    Europe    Eastern Europe         Poland  not available in demo dataset   
2  Americas     South America      Argentina                   Buenos Aires   
3  Americas  Northern America  United States  not available in demo dataset   
4  Americas  Northern America  United States  not available in demo dataset   

                           metro                           city  \
0                      (not set)                      Singapore   
1  not available in demo dataset  not available in demo dataset   
2                      (not set)                   Buenos Aires   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

                          cityId       networkDomain  \
0  not available in demo dataset     unknown.unknown   
1  not available in demo dataset            tpnet.pl   
2  not available in demo dataset  phonevision.com.ar   
3  not available in demo dataset         comcast.net   
4  not available in demo dataset             att.net   

                        latitude                      longitude  \
0  not available in demo dataset  not available in demo dataset   
1  not available in demo dataset  not available in demo dataset   
2  not available in demo dataset  not available in demo dataset   
3  not available in demo dataset  not available in demo dataset   
4  not available in demo dataset  not available in demo dataset   

                 networkLocation  
0  not available in demo dataset  
1  not available in demo dataset  
2  not available in demo dataset  
3  not available in demo dataset  
4  not available in demo dataset

In [19]:
trafficSource = pd.DataFrame([json.loads(linha) for linha in df['trafficSource']])

In [20]:
trafficSource.head()

campaign    source     medium         keyword  \
0         (not set)    google    organic  (not provided)   
1         (not set)    google    organic  (not provided)   
2  Data Share Promo  Partners  affiliate             NaN   
3         (not set)    google    organic  (not provided)   
4         (not set)    google    organic  (not provided)   

                                    adwordsClickInfo isTrueDirect  \
0  {'criteriaParameters': 'not available in demo ...          NaN   
1  {'criteriaParameters': 'not available in demo ...          NaN   
2  {'criteriaParameters': 'not available in demo ...          NaN   
3  {'criteriaParameters': 'not available in demo ...         True   
4  {'criteriaParameters': 'not available in demo ...          NaN   

  referralPath adContent campaignCode  
0          NaN       NaN          NaN  
1          NaN       NaN          NaN  
2          NaN       NaN          NaN  
3          NaN       NaN          NaN  
4          NaN       NaN          NaN

In [21]:
totals = pd.DataFrame([json.loads(linha) for linha in df['totals']])

In [22]:
totals.head()

visits hits pageviews bounces newVisits transactionRevenue
0      1    1         1       1         1                NaN
1      1    1         1       1         1                NaN
2      1    1         1       1         1                NaN
3      1    1         1       1       NaN                NaN
4      1    5         4     NaN         1                NaN

In [23]:
df = df.join(device, how = 'left', lsuffix = '_left', rsuffix = '_right')
df

channelGrouping      date  \
0      Organic Search  20160902   
1      Organic Search  20160902   
2          Affiliates  20160902   
3      Organic Search  20160902   
4      Organic Search  20160902   
...               ...       ...   
12278          Social  20170104   
12279          Social  20170104   
12280          Social  20170104   
12281          Social  20170104   
12282          Social  20170104   

                                                  device        fullVisitorId  \
0      {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1      {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2      {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
...                                                  ...                  ...   
12278  {"browser": "Chrome", "browserVersion": "not a...  7330357551242339916   
12279  {"browser": "Chrome", "browserVersion": "not a...   696320405243257438   
12280  {"browser": "Chrome", "browserVersion": "not a...  5437703324867866371   
12281  {"browser": "Chrome", "browserVersion": "not a...  8869363093179043393   
12282  {"browser": "Android Webview", "browserVersion...  2905034275926274980   

                                              geoNetwork  \
0      {"continent": "Asia", "subContinent": "Southea...   
1      {"continent": "Europe", "subContinent": "Easte...   
2      {"continent": "Americas", "subContinent": "Sou...   
3      {"continent": "Americas", "subContinent": "Nor...   
4      {"continent": "Americas", "subContinent": "Nor...   
...                                                  ...   
12278  {"continent": "Europe", "subContinent": "Easte...   
12279  {"continent": "Europe", "subContinent": "Weste...   
12280  {"continent": "Asia", "subContinent": "Souther...   
12281  {"continent": "Europe", "subContinent": "North...   
12282  {"continent": "Africa", "subContinent": "South...   

                            sessionId  socialEngagementType  \
0      9674781571160116268_1472804607  Not Socially Engaged   
1      8590648239310839049_1472835928  Not Socially Engaged   
2      9772828344252850532_1472856802  Not Socially Engaged   
3      1350700416054916432_1472879649  Not Socially Engaged   
4      1350700416054916432_1472829671  Not Socially Engaged   
...                               ...                   ...   
12278  7330357551242339916_1483540608  Not Socially Engaged   
12279  0696320405243257438_1483580407  Not Socially Engaged   
12280  5437703324867866371_1483592122  Not Socially Engaged   
12281  8869363093179043393_1483601927  Not Socially Engaged   
12282  2905034275926274980_1483593737  Not Socially Engaged   

                                                  totals  \
0      {"visits": "1", "hits": "1", "pageviews": "1",...   
1      {"visits": "1", "hits": "1", "pageviews": "1",...   
2      {"visits": "1", "hits": "1", "pageviews": "1",...   
3      {"visits": "1", "hits": "1", "pageviews": "1",...   
4      {"visits": "1", "hits": "5", "pageviews": "4",...   
...                                                  ...   
12278  {"visits": "1", "hits": "1", "pageviews": "1",...   
12279  {"visits": "1", "hits": "1", "pageviews": "1",...   
12280  {"visits": "1", "hits": "1", "pageviews": "1",...   
12281     {"visits": "1", "hits": "2", "pageviews": "2"}   
12282  {"visits": "1", "hits": "3", "pageviews": "3",...   

                                           trafficSource     visitId  ...  \
0      {"campaign": "(not set)", "source": "google", ...  1472804607  ...   
1      {"campaign": "(not set)", "source": "google", ...  1472835928  ...   
2      {"campaign": "Data Share Promo", "source": "Pa...  1472856802  ...   
3      {"campaign": "(not set)", "source": "google", ...  1472879649  ...   
4      {"campaign": "(not se

In [24]:
df = df.join(geoNetwork, how = 'left', lsuffix = '_left', rsuffix = '_right')

In [25]:
df

channelGrouping      date  \
0      Organic Search  20160902   
1      Organic Search  20160902   
2          Affiliates  20160902   
3      Organic Search  20160902   
4      Organic Search  20160902   
...               ...       ...   
12278          Social  20170104   
12279          Social  20170104   
12280          Social  20170104   
12281          Social  20170104   
12282          Social  20170104   

                                                  device        fullVisitorId  \
0      {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1      {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2      {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
...                                                  ...                  ...   
12278  {"browser": "Chrome", "browserVersion": "not a...  7330357551242339916   
12279  {"browser": "Chrome", "browserVersion": "not a...   696320405243257438   
12280  {"browser": "Chrome", "browserVersion": "not a...  5437703324867866371   
12281  {"browser": "Chrome", "browserVersion": "not a...  8869363093179043393   
12282  {"browser": "Android Webview", "browserVersion...  2905034275926274980   

                                              geoNetwork  \
0      {"continent": "Asia", "subContinent": "Southea...   
1      {"continent": "Europe", "subContinent": "Easte...   
2      {"continent": "Americas", "subContinent": "Sou...   
3      {"continent": "Americas", "subContinent": "Nor...   
4      {"continent": "Americas", "subContinent": "Nor...   
...                                                  ...   
12278  {"continent": "Europe", "subContinent": "Easte...   
12279  {"continent": "Europe", "subContinent": "Weste...   
12280  {"continent": "Asia", "subContinent": "Souther...   
12281  {"continent": "Europe", "subContinent": "North...   
12282  {"continent": "Africa", "subContinent": "South...   

                            sessionId  socialEngagementType  \
0      9674781571160116268_1472804607  Not Socially Engaged   
1      8590648239310839049_1472835928  Not Socially Engaged   
2      9772828344252850532_1472856802  Not Socially Engaged   
3      1350700416054916432_1472879649  Not Socially Engaged   
4      1350700416054916432_1472829671  Not Socially Engaged   
...                               ...                   ...   
12278  7330357551242339916_1483540608  Not Socially Engaged   
12279  0696320405243257438_1483580407  Not Socially Engaged   
12280  5437703324867866371_1483592122  Not Socially Engaged   
12281  8869363093179043393_1483601927  Not Socially Engaged   
12282  2905034275926274980_1483593737  Not Socially Engaged   

                                                  totals  \
0      {"visits": "1", "hits": "1", "pageviews": "1",...   
1      {"visits": "1", "hits": "1", "pageviews": "1",...   
2      {"visits": "1", "hits": "1", "pageviews": "1",...   
3      {"visits": "1", "hits": "1", "pageviews": "1",...   
4      {"visits": "1", "hits": "5", "pageviews": "4",...   
...                                                  ...   
12278  {"visits": "1", "hits": "1", "pageviews": "1",...   
12279  {"visits": "1", "hits": "1", "pageviews": "1",...   
12280  {"visits": "1", "hits": "1", "pageviews": "1",...   
12281     {"visits": "1", "hits": "2", "pageviews": "2"}   
12282  {"visits": "1", "hits": "3", "pageviews": "3",...   

                                           trafficSource     visitId  ...  \
0      {"campaign": "(not set)", "source": "google", ...  1472804607  ...   
1      {"campaign": "(not set)", "source": "google", ...  1472835928  ...   
2      {"campaign": "Data Share Promo", "source": "Pa...  1472856802  ...   
3      {"campaign": "(not set)", "source": "google", ...  1472879649  ...   
4      {"campaign": "(not se

In [26]:
df = df.join(trafficSource, how = 'left', lsuffix = '_left', rsuffix = '_right')

In [27]:
df

channelGrouping      date  \
0      Organic Search  20160902   
1      Organic Search  20160902   
2          Affiliates  20160902   
3      Organic Search  20160902   
4      Organic Search  20160902   
...               ...       ...   
12278          Social  20170104   
12279          Social  20170104   
12280          Social  20170104   
12281          Social  20170104   
12282          Social  20170104   

                                                  device        fullVisitorId  \
0      {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1      {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2      {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4      {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
...                                                  ...                  ...   
12278  {"browser": "Chrome", "browserVersion": "not a...  7330357551242339916   
12279  {"browser": "Chrome", "browserVersion": "not a...   696320405243257438   
12280  {"browser": "Chrome", "browserVersion": "not a...  5437703324867866371   
12281  {"browser": "Chrome", "browserVersion": "not a...  8869363093179043393   
12282  {"browser": "Android Webview", "browserVersion...  2905034275926274980   

                                              geoNetwork  \
0      {"continent": "Asia", "subContinent": "Southea...   
1      {"continent": "Europe", "subContinent": "Easte...   
2      {"continent": "Americas", "subContinent": "Sou...   
3      {"continent": "Americas", "subContinent": "Nor...   
4      {"continent": "Americas", "subContinent": "Nor...   
...                                                  ...   
12278  {"continent": "Europe", "subContinent": "Easte...   
12279  {"continent": "Europe", "subContinent": "Weste...   
12280  {"continent": "Asia", "subContinent": "Souther...   
12281  {"continent": "Europe", "subContinent": "North...   
12282  {"continent": "Africa", "subContinent": "South...   

                            sessionId  socialEngagementType  \
0      9674781571160116268_1472804607  Not Socially Engaged   
1      8590648239310839049_1472835928  Not Socially Engaged   
2      9772828344252850532_1472856802  Not Socially Engaged   
3      1350700416054916432_1472879649  Not Socially Engaged   
4      1350700416054916432_1472829671  Not Socially Engaged   
...                               ...                   ...   
12278  7330357551242339916_1483540608  Not Socially Engaged   
12279  0696320405243257438_1483580407  Not Socially Engaged   
12280  5437703324867866371_1483592122  Not Socially Engaged   
12281  8869363093179043393_1483601927  Not Socially Engaged   
12282  2905034275926274980_1483593737  Not Socially Engaged   

                                                  totals  \
0      {"visits": "1", "hits": "1", "pageviews": "1",...   
1      {"visits": "1", "hits": "1", "pageviews": "1",...   
2      {"visits": "1", "hits": "1", "pageviews": "1",...   
3      {"visits": "1", "hits": "1", "pageviews": "1",...   
4      {"visits": "1", "hits": "5", "pageviews": "4",...   
...                                                  ...   
12278  {"visits": "1", "hits": "1", "pageviews": "1",...   
12279  {"visits": "1", "hits": "1", "pageviews": "1",...   
12280  {"visits": "1", "hits": "1", "pageviews": "1",...   
12281     {"visits": "1", "hits": "2", "pageviews": "2"}   
12282  {"visits": "1", "hits": "3", "pageviews": "3",...   

                                           trafficSource     visitId  ...  \
0      {"campaign": "(not set)", "source": "google", ...  1472804607  ...   
1      {"campaign": "(not set)", "source": "google", ...  1472835928  ...   
2      {"campaign": "Data Share Promo", "source": "Pa...  1472856802  ...   
3      {"campaign": "(not set)", "source": "google", ...  1472879649  ...   
4      {"campaign": "(not se

In [28]:
df = df.join(totals, how = 'left', lsuffix = '_left', rsuffix = '_right')

In [29]:
df.head()

channelGrouping      date  \
0  Organic Search  20160902   
1  Organic Search  20160902   
2      Affiliates  20160902   
3  Organic Search  20160902   
4  Organic Search  20160902   

                                              device        fullVisitorId  \
0  {"browser": "Chrome", "browserVersion": "not a...  9674781571160116268   
1  {"browser": "Chrome", "browserVersion": "not a...  8590648239310839049   
2  {"browser": "Chrome", "browserVersion": "not a...  9772828344252850532   
3  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   
4  {"browser": "Safari", "browserVersion": "not a...  1350700416054916432   

                                          geoNetwork  \
0  {"continent": "Asia", "subContinent": "Southea...   
1  {"continent": "Europe", "subContinent": "Easte...   
2  {"continent": "Americas", "subContinent": "Sou...   
3  {"continent": "Americas", "subContinent": "Nor...   
4  {"continent": "Americas", "subContinent": "Nor...   

                        sessionId  socialEngagementType  \
0  9674781571160116268_1472804607  Not Socially Engaged   
1  8590648239310839049_1472835928  Not Socially Engaged   
2  9772828344252850532_1472856802  Not Socially Engaged   
3  1350700416054916432_1472879649  Not Socially Engaged   
4  1350700416054916432_1472829671  Not Socially Engaged   

                                              totals  \
0  {"visits": "1", "hits": "1", "pageviews": "1",...   
1  {"visits": "1", "hits": "1", "pageviews": "1",...   
2  {"visits": "1", "hits": "1", "pageviews": "1",...   
3  {"visits": "1", "hits": "1", "pageviews": "1",...   
4  {"visits": "1", "hits": "5", "pageviews": "4",...   

                                       trafficSource     visitId  ...  \
0  {"campaign": "(not set)", "source": "google", ...  1472804607  ...   
1  {"campaign": "(not set)", "source": "google", ...  1472835928  ...   
2  {"campaign": "Data Share Promo", "source": "Pa...  1472856802  ...   
3  {"campaign": "(not set)", "source": "google", ...  1472879649  ...   
4  {"campaign": "(not set)", "source": "google", ...  1472829671  ...   

   isTrueDirect  referralPath adContent campaignCode visits hits pageviews  \
0           NaN           NaN       NaN          NaN      1    1         1   
1           NaN           NaN       NaN          NaN      1    1         1   
2           NaN           NaN       NaN          NaN      1    1         1   
3          True           NaN       NaN          NaN      1    1         1   
4           NaN           NaN       NaN          NaN      1    5         4   

   bounces newVisits transactionRevenue  
0        1         1                NaN  
1        1         1                NaN  
2        1         1                NaN  
3        1       NaN                NaN  
4      NaN         1                NaN  

[5 rows x 54 columns]

In [30]:
df.drop(columns = ['device', 'geoNetwork', 'totals', 'trafficSource'], axis = 1, inplace = True)

In [31]:
df

channelGrouping      date        fullVisitorId  \
0      Organic Search  20160902  9674781571160116268   
1      Organic Search  20160902  8590648239310839049   
2          Affiliates  20160902  9772828344252850532   
3      Organic Search  20160902  1350700416054916432   
4      Organic Search  20160902  1350700416054916432   
...               ...       ...                  ...   
12278          Social  20170104  7330357551242339916   
12279          Social  20170104   696320405243257438   
12280          Social  20170104  5437703324867866371   
12281          Social  20170104  8869363093179043393   
12282          Social  20170104  2905034275926274980   

                            sessionId  socialEngagementType     visitId  \
0      9674781571160116268_1472804607  Not Socially Engaged  1472804607   
1      8590648239310839049_1472835928  Not Socially Engaged  1472835928   
2      9772828344252850532_1472856802  Not Socially Engaged  1472856802   
3      1350700416054916432_1472879649  Not Socially Engaged  1472879649   
4      1350700416054916432_1472829671  Not Socially Engaged  1472829671   
...                               ...                   ...         ...   
12278  7330357551242339916_1483540608  Not Socially Engaged  1483540608   
12279  0696320405243257438_1483580407  Not Socially Engaged  1483580407   
12280  5437703324867866371_1483592122  Not Socially Engaged  1483592122   
12281  8869363093179043393_1483601927  Not Socially Engaged  1483601927   
12282  2905034275926274980_1483593737  Not Socially Engaged  1483593737   

       visitNumber  visitStartTime          browser  \
0                1      1472804607           Chrome   
1                1      1472835928           Chrome   
2                1      1472856802           Chrome   
3                2      1472879649           Safari   
4                1      1472829671           Safari   
...            ...             ...              ...   
12278            1      1483540608           Chrome   
12279            1      1483580407           Chrome   
12280            1      1483592122           Chrome   
12281            2      1483601927           Chrome   
12282            1      1483593737  Android Webview   

                      browserVersion  ... isTrueDirect   referralPath  \
0      not available in demo dataset  ...          NaN            NaN   
1      not available in demo dataset  ...          NaN            NaN   
2      not available in demo dataset  ...          NaN            NaN   
3      not available in demo dataset  ...         True            NaN   
4      not available in demo dataset  ...          NaN            NaN   
...                              ...  ...          ...            ...   
12278  not available in demo dataset  ...          NaN  /yt/about/pl/   
12279  not available in demo dataset  ...          NaN     /yt/about/   
12280  not available in demo dataset  ...          NaN     /yt/about/   
12281  not available in demo dataset  ...         True  /yt/about/lt/   
12282  not available in demo dataset  ...          NaN     /yt/about/   

      adContent  campaignCode visits hits pageviews bounces newVisits  \
0           NaN           NaN      1    1         1       1         1   
1           NaN           NaN      1    1         1       1         1   
2           NaN           NaN      1    1         1       1         1   
3           NaN           NaN      1    1         1       1       NaN   
4           NaN           NaN      1    5         4     NaN         1   
...         ...           ...    ...  ...       ...     ...       ...   
12278       NaN           NaN      1    1         1       1         1   
12279       NaN           NaN      1    1         1       1         1   
12280       NaN           NaN      1    1         1       1         1   
12281       NaN           NaN      1    2         2     NaN       NaN   
12282       NaN           NaN      1    3         3     NaN         1   

      transaction

# Limpando a base de dados

É importante analisar a quantidade de valores distintos há em cada coluna para passar ao modelo.

In [32]:
len(df['channelGrouping'].unique())

8

In [35]:
for coluna in df.columns:
    print(coluna +': ' + str(len(df[coluna].unique())))

channelGrouping: 8
date: 366
fullVisitorId: 9997
sessionId: 12276
socialEngagementType: 1
visitId: 12272
visitNumber: 50
visitStartTime: 12279
browser: 22
browserVersion: 1
browserSize: 1
operatingSystem: 12
operatingSystemVersion: 1
isMobile: 2
mobileDeviceBranding: 1
mobileDeviceModel: 1
mobileInputSelector: 1
mobileDeviceInfo: 1
mobileDeviceMarketingName: 1
flashVersion: 1
language: 1
screenColors: 1
screenResolution: 1
deviceCategory: 3
continent: 6
subContinent: 22
country: 145
region: 212
metro: 53
city: 296
cityId: 1
networkDomain: 1747
latitude: 1
longitude: 1
networkLocation: 1
campaign: 7
source: 76
medium: 7
keyword: 129


TypeError: unhashable type: 'dict'

In [36]:
coluna

'adwordsClickInfo'

In [37]:
df['adwordsClickInfo']

0        {'criteriaParameters': 'not available in demo ...
1        {'criteriaParameters': 'not available in demo ...
2        {'criteriaParameters': 'not available in demo ...
3        {'criteriaParameters': 'not available in demo ...
4        {'criteriaParameters': 'not available in demo ...
                               ...                        
12278    {'criteriaParameters': 'not available in demo ...
12279    {'criteriaParameters': 'not available in demo ...
12280    {'criteriaParameters': 'not available in demo ...
12281    {'criteriaParameters': 'not available in demo ...
12282    {'criteriaParameters': 'not available in demo ...
Name: adwordsClickInfo, Length: 12283, dtype: object

Há diversos valores do tipo dicionário em cada linha. Avaliemos as informações contidas nesta coluna. 

In [38]:
df['adwordsClickInfo'][0]

{'criteriaParameters': 'not available in demo dataset'}

not available in demo dataset aparece por conta do Google proteger os dados mais sensíveis dos seus usuários. Como em todas as linhas esta informação não trás novos insights por não estar disponível, será necessário droppá-la.

In [39]:
# excluíndo coluna
df.drop(columns = ['adwordsClickInfo'], axis = 1, inplace = True)

In [45]:
# iterando novamente
for coluna in df.columns:
    print(coluna +': ' + str(df[coluna].nunique()))

channelGrouping: 8
date: 366
fullVisitorId: 9997
sessionId: 12276
socialEngagementType: 1
visitId: 12272
visitNumber: 50
visitStartTime: 12279
browser: 22
browserVersion: 1
browserSize: 1
operatingSystem: 12
operatingSystemVersion: 1
isMobile: 2
mobileDeviceBranding: 1
mobileDeviceModel: 1
mobileInputSelector: 1
mobileDeviceInfo: 1
mobileDeviceMarketingName: 1
flashVersion: 1
language: 1
screenColors: 1
screenResolution: 1
deviceCategory: 3
continent: 6
subContinent: 22
country: 145
region: 212
metro: 53
city: 296
cityId: 1
networkDomain: 1747
latitude: 1
longitude: 1
networkLocation: 1
campaign: 7
source: 76
medium: 7
keyword: 128
isTrueDirect: 1
referralPath: 197
adContent: 15
campaignCode: 1
visits: 1
hits: 96
pageviews: 75
bounces: 1
newVisits: 1
transactionRevenue: 146


Colunas que possuem o mesmo valor para todas as linhas não trazem nenhuma informação relevante ao modelo pois se estas possuírem o mesmo valor, isso não trará informação alguma sobre os padrões de compra dos clientes pois não haverá variação destes valores. Logo, será necessário removê-las do dataset.

In [54]:
# lista vazia para colocar colunas com baixa variação
coluna_na = []
for coluna in df.columns:
    if len(df[coluna].unique()) == 1:
        coluna_na.append(coluna)

In [55]:
coluna_na

['socialEngagementType',
 'browserVersion',
 'browserSize',
 'operatingSystemVersion',
 'mobileDeviceBranding',
 'mobileDeviceModel',
 'mobileInputSelector',
 'mobileDeviceInfo',
 'mobileDeviceMarketingName',
 'flashVersion',
 'language',
 'screenColors',
 'screenResolution',
 'cityId',
 'latitude',
 'longitude',
 'networkLocation',
 'visits']

In [56]:
# excluindo colunas
df.drop(coluna_na, axis = 1, inplace = True)

In [57]:
df

channelGrouping      date        fullVisitorId  \
0      Organic Search  20160902  9674781571160116268   
1      Organic Search  20160902  8590648239310839049   
2          Affiliates  20160902  9772828344252850532   
3      Organic Search  20160902  1350700416054916432   
4      Organic Search  20160902  1350700416054916432   
...               ...       ...                  ...   
12278          Social  20170104  7330357551242339916   
12279          Social  20170104   696320405243257438   
12280          Social  20170104  5437703324867866371   
12281          Social  20170104  8869363093179043393   
12282          Social  20170104  2905034275926274980   

                            sessionId     visitId  visitNumber  \
0      9674781571160116268_1472804607  1472804607            1   
1      8590648239310839049_1472835928  1472835928            1   
2      9772828344252850532_1472856802  1472856802            1   
3      1350700416054916432_1472879649  1472879649            2   
4      1350700416054916432_1472829671  1472829671            1   
...                               ...         ...          ...   
12278  7330357551242339916_1483540608  1483540608            1   
12279  0696320405243257438_1483580407  1483580407            1   
12280  5437703324867866371_1483592122  1483592122            1   
12281  8869363093179043393_1483601927  1483601927            2   
12282  2905034275926274980_1483593737  1483593737            1   

       visitStartTime          browser operatingSystem  isMobile  ...  \
0          1472804607           Chrome         Windows     False  ...   
1          1472835928           Chrome       Macintosh     False  ...   
2          1472856802           Chrome       Macintosh     False  ...   
3          1472879649           Safari             iOS      True  ...   
4          1472829671           Safari             iOS      True  ...   
...               ...              ...             ...       ...  ...   
12278      1483540608           Chrome         Windows     False  ...   
12279      1483580407           Chrome         Android      True  ...   
12280      1483592122           Chrome           Linux     False  ...   
12281      1483601927           Chrome         Android      True  ...   
12282      1483593737  Android Webview         Android      True  ...   

              keyword isTrueDirect   referralPath adContent campaignCode hits  \
0      (not provided)          NaN            NaN       NaN          NaN    1   
1      (not provided)          NaN            NaN       NaN          NaN    1   
2                 NaN          NaN            NaN       NaN          NaN    1   
3      (not provided)         True            NaN       NaN          NaN    1   
4      (not provided)          NaN            NaN       NaN          NaN    5   
...               ...          ...            ...       ...          ...  ...   
12278             NaN          NaN  /yt/about/pl/       NaN          NaN    1   
12279             NaN          NaN     /yt/about/       NaN          NaN    1   
12280             NaN          NaN     /yt/about/       NaN          NaN    1   
12281             NaN         True  /yt/about/lt/       NaN          NaN    2   
12282             NaN          NaN     /yt/about/       NaN          NaN    3   

      pageviews bounces newVisits transactionRevenue  
0             1       1         1                NaN  
1             1       1         1                NaN  
2             1       1         1                NaN  
3             1       1       NaN                NaN  
4             4     NaN         1                NaN  
...         ...     ...       ...                ...  
12278         1       1         1                NaN  
12279         1       1         1                NaN  
12280         1       1         1                NaN  
12281         2     NaN       NaN                NaN  
12282         3     NaN         1                NaN  

[12283 rows x 31 columns]

# Feature Engineering

## Entendendo as variáveis

Será necessário criar variáveis que indiquem a usabilidade dos usuários na plataforma de e-commerce para descobrir níveis de interação, clicks, visitas e etc.

As variáveis quantitativas que mais demonstram o engajamento dos usuários serão 4:

* newVisits: coluna dummy que mostra se é a primeira visita do usuário ou não.

* pageviews: coluna que mostra quantas páginas o usuário viu durante a sessão a partir do momento em que ele acessou o site.

* bounces (ou rejeição): Visita que não gera nenhuma interação.

* hits: mede o volume de interação que o Google recebeu do site. Ajuda a entender quanto o usuário interagiu no site.

In [58]:
df.dtypes

channelGrouping       object
date                  object
fullVisitorId         object
sessionId             object
visitId               object
visitNumber            int64
visitStartTime         int64
browser               object
operatingSystem       object
isMobile                bool
deviceCategory        object
continent             object
subContinent          object
country               object
region                object
metro                 object
city                  object
networkDomain         object
campaign              object
source                object
medium                object
keyword               object
isTrueDirect          object
referralPath          object
adContent             object
campaignCode          object
hits                  object
pageviews             object
bounces               object
newVisits             object
transactionRevenue    object
dtype: object

In [60]:
quant = ['bounces', 'hits', 'newVisits', 'pageviews', 'transactionRevenue']

In [61]:
# substitui o valor numérico da coluna
for coluna in quant:
    df[coluna] = pd.to_numeric(df[coluna])

A função to_numeric converte as colunas passadas e identifica o valor e formato mais adequado para cada coluna.

In [62]:
df[quant]

bounces  hits  newVisits  pageviews  transactionRevenue
0          1.0     1        1.0          1                 NaN
1          1.0     1        1.0          1                 NaN
2          1.0     1        1.0          1                 NaN
3          1.0     1        NaN          1                 NaN
4          NaN     5        1.0          4                 NaN
...        ...   ...        ...        ...                 ...
12278      1.0     1        1.0          1                 NaN
12279      1.0     1        1.0          1                 NaN
12280      1.0     1        1.0          1                 NaN
12281      NaN     2        NaN          2                 NaN
12282      NaN     3        1.0          3                 NaN

[12283 rows x 5 columns]

In [63]:
df[quant].dtypes

bounces               float64
hits                    int64
newVisits             float64
pageviews               int64
transactionRevenue    float64
dtype: object

# Missing values da coluna transactionRevenue

In [68]:
df['transactionRevenue']

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
12278   NaN
12279   NaN
12280   NaN
12281   NaN
12282   NaN
Name: transactionRevenue, Length: 12283, dtype: float64

In [69]:
df['transactionRevenue'].value_counts()

13590000.0     3
19190000.0     3
33590000.0     3
31490000.0     2
69990000.0     2
              ..
22490000.0     1
74850000.0     1
32490000.0     1
239880000.0    1
25240000.0     1
Name: transactionRevenue, Length: 146, dtype: int64

In [70]:
df['transactionRevenue'].isnull().sum()

12119

In [71]:
df.shape

(12283, 31)

A grande maioria das colunas do dataframe possui valores nulos.

Alguns modelos de machine learning não aceitarão missing values. Estes valores corresppondem a usuários que não fizeram compras no site. Pode-se transformar estes valores pra zero.

In [72]:
# fillando valores nulos com 0
df['transactionRevenue'].fillna(value = 0, inplace = True)

In [73]:
df['transactionRevenue']

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
12278    0.0
12279    0.0
12280    0.0
12281    0.0
12282    0.0
Name: transactionRevenue, Length: 12283, dtype: float64

Valores nulos agora encontram-se como 0.

In [74]:
df['transactionRevenue'].value_counts()

0.0           12119
13590000.0        3
33590000.0        3
19190000.0        3
27190000.0        2
              ...  
24730000.0        1
6990000.0         1
14480000.0        1
95180000.0        1
25240000.0        1
Name: transactionRevenue, Length: 147, dtype: int64

Os números de gasto são muito grandes pois tratando-se de uma loja de brindes, cada item custam valores baixos.

Todos os valores de transactionRevenue estão multiplicados por 1 milhão segundo a [documentação do Google](https://developers.google.com/analytics/devguides/reporting/core/v4/changelog).

Portanto será necessário corrigir tais valores de transações.

In [75]:
df['transactionRevenue'] = df['transactionRevenue'] / 1000000

In [77]:
df['transactionRevenue'].value_counts()

0.00     12119
13.59        3
33.59        3
19.19        3
27.19        2
         ...  
24.73        1
6.99         1
14.48        1
95.18        1
25.24        1
Name: transactionRevenue, Length: 147, dtype: int64

## Criando Variáveis

Cada usuário possui um fullVisitorId.

In [79]:
len(set(df['fullVisitorId']))

9997

In [80]:
df.shape

(12283, 31)

Alguns usuários utilizaram o site mais de uma vez, pois o número de usuários (9997) é menor que o número total de visitas registradas no dataset.

In [82]:
# soma de todas as colunas quantitativas para cada usuário
df.groupby(by = 'fullVisitorId').sum()

visitNumber  visitStartTime  isMobile  hits  pageviews  \
fullVisitorId                                                                 
0002365800130207040            1      1472974804         0     1          1   
0010286039787739137            1      1475084026         0     4          3   
0011056874471185769            1      1480996024         0     2          2   
0014443856125569702            1      1475423502         0     1          1   
0017260116665815114            6      1492707286         0     2          2   
...                          ...             ...       ...   ...        ...   
9996691020299725061            3      1482388264         0    10         10   
9996793353803109760            1      1480566609         1     1          1   
9997125062030599880            1      1484217422         1     8          8   
9997464307347866920            1      1489504370         1     1          1   
9999355589118867468            3      2992090780         0     2          2   

                     bounces  newVisits  transactionRevenue  
fullVisitorId                                                
0002365800130207040      1.0        1.0                 0.0  
0010286039787739137      0.0        1.0                 0.0  
0011056874471185769      0.0        1.0                 0.0  
0014443856125569702      1.0        1.0                 0.0  
0017260116665815114      0.0        0.0                 0.0  
...                      ...        ...                 ...  
9996691020299725061      0.0        0.0                 0.0  
9996793353803109760      1.0        1.0                 0.0  
9997125062030599880      0.0        1.0                 0.0  
9997464307347866920      1.0        1.0                 0.0  
9999355589118867468      2.0        1.0                 0.0  

[9997 rows x 8 columns]

In [85]:
df_quant = df.groupby(by = 'fullVisitorId', as_index = False)[quant].sum()

In [86]:
df_quant.head()

fullVisitorId  bounces  hits  newVisits  pageviews  \
0  0002365800130207040      1.0     1        1.0          1   
1  0010286039787739137      0.0     4        1.0          3   
2  0011056874471185769      0.0     2        1.0          2   
3  0014443856125569702      1.0     1        1.0          1   
4  0017260116665815114      0.0     2        0.0          2   

   transactionRevenue  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

Agora temos para cada usuário as variáveis quantitativas escolhidas.

In [87]:
df_quant.shape

(9997, 6)

Este será o dataset preparado para criar o modelo de machine learning preditivo.

# Treinando o modelo

In [88]:
# utilizando uma cópia da variável target
y = df_quant[['transactionRevenue']].copy()

# variáveis que serão utilizadas para a predição
X = df_quant.drop(columns = 'transactionRevenue', axis = 1)

In [92]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3,
                                                    random_state = 42)

Será reservado 30% dos dados totais para o teste do modelo.

## Regressão Linear

In [98]:
reg = LinearRegression() # instanciando o modelo

In [100]:
reg.fit(X = X_train, y = y_train) # treinando o modelo com os dados de teste

LinearRegression()

In [103]:
reg_predict = reg.predict(X = X_test) # array com valores preditos

## Resultados da Regressão Linear

In [109]:
resultados = pd.DataFrame() # dataframe de resultados
resultados['revenue'] = y_test # valores comprados pelo usuário
resultados['predict'] = reg_predict # valores preditos
resultados['erro'] = reg_predict - y_test # erro entre a predição e o valor real

In [113]:
resultados.head()

revenue   predict      erro
4122      0.0  1.551660  1.551660
4065      0.0  1.555209  1.555209
1731      0.0  1.687726  1.687726
4740      0.0  1.515492  1.515492
6391      0.0  1.422655  1.422655

In [116]:
# predições para usuários que gastaram
resultados[resultados['revenue'] > 0].head(10)

revenue   predict        erro
3006    32.49  1.616211  -30.873789
5928    74.85  1.449755  -73.400245
8361   424.50  1.306194 -423.193806
8612    55.99  1.291322  -54.698678
416     15.19  1.831725  -13.358275
4077     5.97  1.554336   -4.415664
8146    27.19  1.318915  -25.871085
4908   372.65  1.505642 -371.144358
9395    29.98  1.779852  -28.200148
8728   101.95  1.284318 -100.665682

## Mean square error

In [124]:
MSE = np.mean((reg_predict - y_test) ** 2)
MSE

c:\Users\pedro\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\core\fromnumeric.py:3430: FutureWarning: In a future version, DataFrame.mean(axis=None) will return a scalar mean over the entire DataFrame. To retain the old behavior, use 'frame.mean(axis=0)' or just 'frame.mean()'
  return mean(axis=axis, dtype=dtype, out=out, **kwargs)


transactionRevenue    807.24593
dtype: float64

## Square root mean square error

In [126]:
RSME = np.sqrt(np.mean((reg_predict - y_test) ** 2))
RSME

c:\Users\pedro\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\core\fromnumeric.py:3430: FutureWarning: In a future version, DataFrame.mean(axis=None) will return a scalar mean over the entire DataFrame. To retain the old behavior, use 'frame.mean(axis=0)' or just 'frame.mean()'
  return mean(axis=axis, dtype=dtype, out=out, **kwargs)


transactionRevenue    28.412074
dtype: float64

## RSME no Sci-kit Learn

In [133]:
np.sqrt(mean_squared_error(y_test, reg_predict))

28.412073660366516

In [134]:
np.mean(df_quant['transactionRevenue']) # média de gastos 

1.778953686105832

In [136]:
np.std(df_quant['transactionRevenue']) # desvio padrão dos gastos

25.129905708016967

O modelo está errando mais que o desvio padrão da base original.

# Análise gráfica dos resultados